sw/host/opentitanlib/src/transport/hyperdebug/spi.rs

```diff
@@ -0,0 +1,213 @@
+        let databytes = std::cmp::min(USB_MAX_SIZE - 6, wbuf.len());
+        buf[6..6 + databytes].clone_from_slice(&wbuf[0..databytes]);
+        self.usb_write_bulk(&buf[0..6 + databytes])?;
+        let mut index = databytes;
+
+        while index < wbuf.len() {
+            buf[0..2].clone_from_slice(&USB_SPI_PKT_ID_CMD_TRANSFER_CONTINUE.to_le_bytes());
+            buf[2..4].clone_from_slice(&(index as u16).to_le_bytes());
+            let databytes = std::cmp::min(USB_MAX_SIZE - 4, wbuf.len() - index);
+            buf[4..4 + databytes].clone_from_slice(&wbuf[index..index+databytes]);
+            self.usb_write_bulk(&buf[0..4 + databytes])?;
+            index += databytes;
+        }
+        Ok(())
+    }
+    
+    /// Receive data for a single SPI operation, using one or more USB packets.
+    fn receive(&self, rbuf: &mut [u8]) -> Result<()> {
+        let mut buf: [u8; USB_MAX_SIZE] = [0; USB_MAX_SIZE];
+        let bytecount = self.usb_read_bulk(&mut buf)?;
+        ensure!(bytecount >= 4,
+                Error::CommunicationError("Unrecognized reponse to TRANSFER_START"));
+        ensure!(buf[0..2] == USB_SPI_PKT_ID_RSP_TRANSFER_START.to_le_bytes(),
+                Error::CommunicationError("Unrecognized reponse to TRANSFER_START"));
+        let status_code = u16::from_le_bytes(buf[2..4].try_into()?);
+        ensure!(status_code == 0, Error::CommunicationError("SPI error"));
+        let databytes = bytecount - 4;
+        rbuf[0..databytes].clone_from_slice(&buf[4..4+databytes]);
+        let mut index = databytes;
+        while index < rbuf.len() {
+            let bytecount = self.usb_read_bulk(&mut buf)?;
+            ensure!(bytecount > 4,
+                    Error::CommunicationError("Unrecognized reponse to TRANSFER_START"));
+            ensure!(buf[0..2] == USB_SPI_PKT_ID_RSP_TRANSFER_CONTINUE.to_le_bytes(),
+                    Error::CommunicationError("Unrecognized reponse to TRANSFER_START"));
+            ensure!(buf[2..4] == (index as u16).to_le_bytes(),
+                    Error::CommunicationError("Unexpected byte index in reponse to TRANSFER_START"));
+            let databytes = bytecount - 4;
+            rbuf[index..index+databytes].clone_from_slice(&buf[4..4+databytes]);
+            index += databytes;
+        }
+        Ok(())
+    }
+
+    /// Send one USB packet.
+    fn usb_write_bulk(&self, buf: &[u8]) -> Result<()> {
+        self.inner.usb_device.borrow().write_bulk(self.interface.out_endpoint, buf)?;
+        Ok(())
+    }
+
+    /// Receive one USB packet.
+    fn usb_read_bulk(&self, buf: &mut [u8]) -> Result<usize> {
+        Ok(self.inner.usb_device.borrow().read_bulk(self.interface.in_endpoint, buf)?)
+    }
+}
+
+impl Target for HyperdebugSpiTarget {
+    fn get_transfer_mode(&self) -> Result<TransferMode> {
+        Ok(TransferMode::Mode0)
+    }
+    fn set_transfer_mode(&self, _mode: TransferMode) -> Result<()> {
+        unimplemented!();
```

Comment:
would `todo!()` be more appropriate?

Reply:
Yes, probably.  I did not know of that macro.